src/org/mozilla/javascript/NativeJavaList.java

```diff
@@ -0,0 +1,196 @@
+/* -*- Mode: java; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.javascript;
+
+
+import java.util.List;
+
+/**
+ * This class reflects Java Lists into the JavaScript environment.
+ *
+ * @author Markus Sunela
+ * @see NativeJavaClass
+ * @see NativeJavaObject
+ * @see NativeJavaPackage
+ */
+
+public class NativeJavaList
+        extends NativeJavaObject
+        implements SymbolScriptable {
+
+    private static final long serialVersionUID = -924022554283675333L;
+
+    /**
+     * Implementation of the Array.includes method.
+     */
+    private static class IncludesMethod extends BaseFunction {
+        private List list;
+
+        IncludesMethod(List list) {
+            this.list = list;
+        }
+
+        @Override
+        public Object call(Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
+            Object arg = args[0];
+            if (arg instanceof Wrapper)
+                arg = ((Wrapper)arg).unwrap();
+
+            int start = 0;
+            if (args.length > 1)
+                start = ((Number)args[1]).intValue();
+
+            if (start == 0)
+                return list.contains(arg);
+            else
+                return list.subList(start, list.size()).contains(arg);
+        }
+    }
+
+    @Override
+    public String getClassName() {
+        return "JavaList";
+    }
+
+    public static NativeJavaList wrap(Scriptable scope, Object array) {
+        return new NativeJavaList(scope, array);
+    }
+
+    @Override
+    public Object unwrap() {
+        return list;
+    }
+
+    public NativeJavaList(Scriptable scope, Object list) {
+        super(scope, list, list.getClass());
+        Class<?> cl = list.getClass();
+        if (!List.class.isAssignableFrom(cl)) {
+            throw new RuntimeException("List expected");
+        }
+        this.list = (List) list;
+        this.cls = cl;
+    }
+
+    @Override
+    public boolean has(String id, Scriptable start) {
+        return id.equals("length") || id.equals("includes") || super.has(id, start);
+    }
+
+    @Override
+    public boolean has(int index, Scriptable start) {
+        return 0 <= index && index < list.size();
+    }
+
+    @Override
+    public boolean has(Symbol key, Scriptable start) {
+        return SymbolKey.IS_CONCAT_SPREADABLE.equals(key);
+    }
+
+    @Override
+    public Object get(String id, Scriptable start) {
+        if (id.equals("length"))
+            return Integer.valueOf(list.size());
+        else if (id.equals("includes"))
```

Comment:
This is only one extension of the list.
I think we were both trying to reach the same goal: See https://github.com/mozilla/rhino/pull/830
A `JavaList` should be treated like an `Array` and provide functions like `join`, `includes` etc...